File: TinyScript/js/src/test/commons/PeekIterator.test.ts
import arrayToGenerator from '../../commons/ArrayToGenerator'
import PeekIterator from '../../commons/PeekIterator'


describe("test PeekIterator", () => {

    test('test next', () => {
        const data = 'abcdefg'
        const it = new PeekIterator(arrayToGenerator([...data]))
        expect(it.next()).toEqual('a')
        expect(it.next()).toEqual('b')
        expect(it.next()).toEqual('c')
        expect(it.next()).toEqual('d')
        expect(it.next()).toEqual('e')
        expect(it.next()).toEqual('f')
        expect(it.next()).toEqual('g')
    })

    test('test pee', () => {
        const data = 'abcdefg'
        const it = new PeekIterator(arrayToGenerator([...data]))
        expect(it.next()).toEqual('a')
        it.peek()
        it.peek()
        expect(it.next()).toEqual('b')
    })

    test('test endToken', () => {
        const data = 'abcdefg'
        const it = new PeekIterator(arrayToGenerator([...data]), '\0')
        for (let i = 0; i < 8; i++) {
            if (i === 7) {
                expect(it.next()).toEqual('\0')
            } else {
                it.next()
            }
        }
    })

    test("test peek and next",()=>{
        const data = 'abcdefg'
        const it = new PeekIterator(arrayToGenerator([...data]))
        expect(it.next()).toEqual('a')
        expect(it.next()).toEqual('b')
        expect(it.next()).toEqual('c')
        it.peek();
        expect(it.next()).toEqual('d')
        expect(it.next()).toEqual('e')
    })
})

File: TinyScript/java/src/test/java/commons/PeekIteratorTest.java
package commons;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

/**
 * @author MFine
 * @version 1.0
 * @date 2021/7/1 0:00
 **/
class PeekIteratorTest {
    @Test
    public void test_peek() {
        var source = "abcdefg";
        var it = new PeekIterator<>(source.chars().mapToObj(c -> (char) c));

        assertEquals('a', it.next());
        assertEquals('b', it.next());
        it.next();
        it.next();
        assertEquals('e', it.next());
        assertEquals('f', it.peek());
        assertEquals('f', it.peek());
        assertEquals('f', it.next());
        assertEquals('g', it.next());
    }

    @Test
    public void testLookAhead() {
        var source = "abcdefg";
        var it = new PeekIterator<>(source.chars().mapToObj(c -> (char) c));
        assertEquals('a', it.next());
        assertEquals('b', it.next());
        it.putBack();
        it.putBack();
        assertEquals('a', it.next());
    }

    @Test
    public void testEndToken() {
        var source = "abcdefg";
        var it = new PeekIterator<>(source.chars().mapToObj(c -> (char) c));
        var i = 0;
        while (it.hasNext()) {
            if (i == 7) {
                assertEquals((char) 0, it.next());
            } else {
                assertEquals(source.charAt(i++), it.next());
            }
        }
    }
}